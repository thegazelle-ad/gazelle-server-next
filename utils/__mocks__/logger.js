export const logger = {
  fatal: jest.fn().mockResolvedValue(undefined),
  error: jest.fn(),
  warn: jest.fn(),
  debug: jest.fn(),
};

logger.__mockClear = () => {
  logger.fatal.mockClear();
  logger.error.mockClear();
  logger.warn.mockClear();
  logger.debug.mockClear();
};
