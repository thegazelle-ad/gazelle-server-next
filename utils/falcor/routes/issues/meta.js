import falcor from 'falcor';

import * as db from 'lib/db';

const $ref = falcor.Model.ref;

export default [
  {
    // Get latest issue
    route: "issues['latest']",
    get: () =>
      new Promise(resolve => {
        db.latestIssueQuery().then(row => {
          resolve([
            {
              path: ['issues', 'latest'],
              value: $ref(['issues', 'byNumber', row[0].issue_number]),
            },
          ]);
        });
      }),
  },
];
