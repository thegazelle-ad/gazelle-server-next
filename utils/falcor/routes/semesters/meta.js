import falcor from 'falcor';

import * as db from 'lib/db';

const $ref = falcor.Model.ref;

export default [
  {
    route: "semesters['latest']",
    get: () =>
      new Promise((resolve, reject) => {
        db.getLatestSemester()
          .then(semesterName => {
            resolve([
              {
                path: ['semesters', 'latest'],
                value: $ref(['semesters', 'byName', semesterName]),
              },
            ]);
          })
          .catch(reject);
      }),
  },
];
