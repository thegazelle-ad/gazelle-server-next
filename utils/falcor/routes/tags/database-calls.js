import _ from 'lodash';

import { database } from 'lib/db';

/**
 * @param   {Object[]}   tagObject the tag to be added.
 * @returns {boolean}    tags      number of tags modified
 */
export async function addTag(tagObject) {
  await database('tags').insert(tagObject);
  return true;
}

/**
 * Fetches direct meta data of tags from the tags database table
 * @param {string} queryField - Indicates which field to query by
 * @param {string[]} queryParams - Array of parameters of type queryField of articles to fetch
 * @param {string[]} columns - Which columns of the articles table to fetch
 * @returns {Promise<Object[]>}
 */
export function tagQuery(queryField, queryParams, columns) {
  // In order to be able to identify the rows we get back we need to include the queryField
  if (!columns.includes(queryField)) {
    columns.push(queryField);
  }
  return database
    .select(...columns)
    .from('tags')
    .whereIn(queryField, queryParams);
}

/**
 * Update tags given a json graph.
 * @param   {Object}   jsonGraphArg - the json graph to update
 */
export function updateTags(jsonGraphArg) {
  _.forEach(jsonGraphArg, async (tagObject, slug) => {
    await database('tags')
      .where('slug', '=', slug)
      .update(tagObject)
      .then(data => {
        if (data !== 1) {
          throw new Error(`Problems updating tag with slug: ${slug}`);
        }
      })
      .catch(e => {
        throw e;
      });
  });
}
