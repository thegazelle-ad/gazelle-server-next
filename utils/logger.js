/**
 * This file exposes our logger in a module form. It has the private variables
 * isClient and displayAlert that can only be accessed through the exposed API
 * of initializeLogger and updateDisplayAlert. If you come from the object
 * oriented programming paradigm this is equivalent to the singleton pattern,
 * except we never even created a class so we don't need to bother with making
 * sure it only gets instantiated once as it is now impossible.
 *
 * the logger object is the actual logger to be imported, but before it is used
 * initializeLogger should be called somewhere in the global scope, and if on the
 * client side and you want to add a prettier displayAlert function that is not necessarily
 * available at the start of the process, you can later call updateDisplayAlert
 * to make this your displayAlert function.
 */
import _ from 'lodash';

let isClient = null;
let displayAlert = null;

function validateInitialization() {
  if (isClient === null || (isClient && !_.isFunction(displayAlert)))
    throw new Error('The logger has yet to be initialized');
}

export const logger = {
  /**
   * Logs a fatal error. This should only be called if an error occurred that
   * completely breaks our system and requires a complete reboot of the code.
   * On the client it will refresh the page after displaying an alert,
   * on the server it will exit the server which will make 'forever' restart
   * the server.
   * @param {Error | string} err - The fatal error to be logged, either a string or
   * an Error object
   */
  async fatal(err) {
    validateInitialization();

    const errorObject = err instanceof Error ? err : new Error(err);
    if (isClient) {
      // We are actually lying currently, as developers haven't been notified
      // when we hook up sentry we will be notifying developers though
      await displayAlert(
        'A fatal error has occurred, and the developers' +
          ' have been notified. We apologize for the inconvenience, and we are afraid' +
          ' that we will have to reload the page now',
      );
      window.location.reload(true);
    } else {
      // eslint-disable-next-line no-console
      console.error(
        'A fatal error occured, and the server is exiting. See the error below:',
      );
      // eslint-disable-next-line no-console
      console.error(errorObject);
      process.exit(1);
    }
  },

  /**
   * Logs an error. This should only be called on unexpected behaviour that should
   * never happen, but it doesn't completely break our system
   * @param {Error | string} err - The error to be logged, either a string or
   * an Error object
   */
  error(err) {
    validateInitialization();

    const errorObject = err instanceof Error ? err : new Error(err);
    // eslint-disable-next-line no-console
    console.error(errorObject);
  },

  /**
   * Logs a warning. This should be called when an event occurs that is usually unwanted
   * but under certain circumstances could be okay
   * @param {Error | string} warning - The warning to be logged, either a string or
   * an Error object
   */
  warn(warning) {
    validateInitialization();

    // eslint-disable-next-line no-console
    console.warn(warning);
  },

  /**
   * Logs a debug statement. This can be called whereever the developer thinks
   * information could be helpful to have when debugging. It is the lowest logging
   * level so will normally only be chosen to be shown in development mode
   * @param {Error | string} msg - The warning to be logged, either a string or
   * an Error object
   */
  debug(msg) {
    validateInitialization();

    // eslint-disable-next-line no-console
    console.log(msg);
  },
};

/**
 * @param {boolean} updatedIsClient - Whether we are currently on the client or the server
 * @param {func} [updatedDisplayAlert] - This argument displays an alert on the client side
 * to the user, and should therefore only be supplied if isClient is true
 */
export const initializeLogger = (updatedIsClient, updatedDisplayAlert) => {
  if (updatedIsClient === undefined) {
    throw new Error('initializeLogger needs an isClient parameter supplied');
  }
  if (updatedIsClient && !_.isFunction(updatedDisplayAlert)) {
    throw new Error(
      'On the client initializeLogger needs a displayAlert function supplied',
    );
  }
  isClient = updatedIsClient;
  displayAlert = updatedDisplayAlert;
};

/**
 * Should only be called on the client, and should be used for updating the alert
 * function to a prettier one that is not initialized before initial render.
 * Expected use case is using window.alert as default and then switching it when
 * React has initialized and a nicer alert function has been created.
 * @param {func} newDisplayAlert - This argument displays an alert on the client side
 * to the user, and should therefore only be supplied if isClient is true
 */
export const updateDisplayAlert = newDisplayAlert => {
  if (!_.isFunction(newDisplayAlert)) {
    throw new Error(
      'updateDisplayAlert needs an alert function passed as a parameter',
    );
  }
  displayAlert = newDisplayAlert;
};
