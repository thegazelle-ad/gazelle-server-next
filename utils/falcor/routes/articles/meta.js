import { getNumArticles, database } from 'lib/db';
import { logger } from 'lib/logger';
import { createNewArticle } from './database-calls.sql';
import { buildErrorMessage } from 'lib/error-helpers';
import { has } from 'lib/utilities';
import falcor from 'falcor';
import _ from 'lodash';

const $ref = falcor.Model.ref;

export const routes = [
  {
    route: "articles['length']",
    get: async () => {
      const length = await getNumArticles();
      return [
        {
          path: ['articles', 'length'],
          value: length,
        },
      ];
    },
  },
  {
    route: "articles['createNew']",
    call: async (callPath, args) => {
      const article = args[0];
      if (!article || !article.title || !article.slug) {
        logger.error(
          new Error(
            "articles['createNew'] must be provided both a slug and a title",
          ),
        );
        throw new Error(buildErrorMessage());
      }
      const createdArticle = await createNewArticle(database, article);
      if (!createdArticle.id) {
        logger.error(
          new Error('New article returned from db must include an id'),
        );
        throw new Error(buildErrorMessage());
      }
      const pathPrefix = ['articles', 'byId', createdArticle.id];
      let processedFields = 0;
      const standardFields = ['id', 'slug', 'title', 'teaser', 'image_url'];
      const results = standardFields
        .map(field => {
          if (has.call(createdArticle, field)) {
            processedFields += 1;
            return {
              path: pathPrefix.concat([field]),
              value: createdArticle[field],
            };
          }
          return null;
        })
        .filter(x => x !== null);
      if (has.call(createdArticle, 'category_id')) {
        processedFields += 1;
        results.push({
          path: pathPrefix.concat(['category']),
          value: $ref(['categories', 'byId', createdArticle.category_id]),
        });
      }
      if (has.call(createdArticle, 'authors')) {
        processedFields += 1;
        results.push(
          ...createdArticle.authors.map((authorRow, index) => ({
            path: pathPrefix.concat(['authors', index]),
            value: $ref(['staff', 'byId', authorRow.author_id]),
          })),
        );
      }
      if (has.call(createdArticle, 'tags')) {
        processedFields += 1;
        results.push(
          ...createdArticle.tags.map((tagRow, index) => ({
            path: pathPrefix.concat(['tags', index]),
            value: $ref(['tags', 'byId', tagRow.tag_id]),
          })),
        );
      }

      if (
        processedFields !==
        Object.keys(_.omit(createdArticle, ['created_at'])).length
      ) {
        logger.warn(
          new Error(
            `Expected every key from the created article to be used\nProcessed fields: ${JSON.stringify(
              processedFields,
            )}\nArticle Fields: ${JSON.stringify(Object.keys(createdArticle))}`,
          ),
        );
      }

      results.push({
        path: ['articles', 'byPage'],
        invalidated: true,
      });
      return results;
    },
  },
];
