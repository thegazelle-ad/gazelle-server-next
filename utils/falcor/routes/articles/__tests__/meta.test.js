import { routes } from '../meta';
import * as dbMethods from '../database-calls.sql';
import { logger } from 'lib/logger';

jest.mock('lib/db');
jest.mock('lib/logger');

const createNewArticlePath = "articles['createNew']";
const createNewArticleRoute = routes.find(
  x => x.route === createNewArticlePath,
);

expect(createNewArticleRoute).not.toBeUndefined();

describe('create new article falcor route', () => {
  dbMethods.createNewArticle = jest.fn().mockResolvedValue({ id: 1 });

  afterEach(() => {
    dbMethods.createNewArticle.mockClear();
    logger.__mockClear();
  });

  it("throws if there isn't an argument", async () => {
    await expect(
      createNewArticleRoute.call(createNewArticlePath, undefined),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('throws if article has no title', async () => {
    await expect(
      createNewArticleRoute.call(createNewArticlePath, [{ slug: 'slug' }]),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('throws if article has no slug', async () => {
    await expect(
      createNewArticleRoute.call(createNewArticlePath, [{ title: 'title' }]),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it('passes article through to the database method', async () => {
    const article = {
      title: 'title',
      slug: 'slug',
    };

    await createNewArticleRoute.call(createNewArticlePath, [article]);
    expect(dbMethods.createNewArticle).toHaveBeenCalledTimes(1);
    expect(dbMethods.createNewArticle.mock.calls[0][1]).toBe(article);
  });

  it('returns expected path sets', async () => {
    dbMethods.createNewArticle.mockResolvedValueOnce({
      authors: [
        {
          article_id: 1,
          author_id: 1,
        },
        {
          article_id: 1,
          author_id: 2,
        },
      ],
      category_id: 3,
      created_at: new Date('2018-04-29T10:48:00.085Z'),
      id: 1,
      image_url: 'https://google.com/img.png',
      slug: 'test-slug',
      tags: [
        {
          article_id: 1,
          tag_id: 5,
        },
        {
          article_id: 1,
          tag_id: 6,
        },
      ],
      teaser: 'test-teaser',
      title: 'test-title',
    });
    await expect(
      createNewArticleRoute.call(createNewArticlePath, [
        { title: 'title', slug: 'slug' },
      ]),
    ).resolves.toMatchSnapshot();
  });
});
