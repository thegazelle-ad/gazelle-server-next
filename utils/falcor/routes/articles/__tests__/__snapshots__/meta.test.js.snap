// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create new article falcor route returns expected path sets 1`] = `
Array [
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "id",
    ],
    "value": 1,
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "slug",
    ],
    "value": "test-slug",
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "title",
    ],
    "value": "test-title",
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "teaser",
    ],
    "value": "test-teaser",
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "image_url",
    ],
    "value": "https://google.com/img.png",
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "category",
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "categories",
        "byId",
        3,
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "authors",
      0,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "staff",
        "byId",
        1,
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "authors",
      1,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "staff",
        "byId",
        2,
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "tags",
      0,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "tags",
        "byId",
        5,
      ],
    },
  },
  Object {
    "path": Array [
      "articles",
      "byId",
      1,
      "tags",
      1,
    ],
    "value": Object {
      "$type": "ref",
      "value": Array [
        "tags",
        "byId",
        6,
      ],
    },
  },
  Object {
    "invalidated": true,
    "path": Array [
      "articles",
      "byPage",
    ],
  },
]
`;

exports[`create new article falcor route throws if article has no slug 1`] = `"An error occured, the developers of The Gazelle have been notified"`;

exports[`create new article falcor route throws if article has no slug 2`] = `
Array [
  Array [
    [Error: articles['createNew'] must be provided both a slug and a title],
  ],
]
`;

exports[`create new article falcor route throws if article has no title 1`] = `"An error occured, the developers of The Gazelle have been notified"`;

exports[`create new article falcor route throws if article has no title 2`] = `
Array [
  Array [
    [Error: articles['createNew'] must be provided both a slug and a title],
  ],
]
`;

exports[`create new article falcor route throws if there isn't an argument 1`] = `"Cannot read property '0' of undefined"`;

exports[`create new article falcor route throws if there isn't an argument 2`] = `Array []`;
